Create DataTree-like type that can be used as a drop in place holder until data types can fully be converted to DataTree

biosafetylvl5 opened the pull request:
*Its transformation ability is perfect. However, if made to laugh, it can’t maintain its disguise.* - Pokemon Scarlet

<img alt="DataTreeDitto" src="https://github.com/user-attachments/assets/cd233926-0a7a-447a-a48f-c9186ede2841" />

Pokemon, [the world's most valuable intellectual property](https://bobrowsky.substack.com/p/how-pokemon-became-the-most-valuable), has a pokemon named "ditto" that duplicates its opponents behavior and structure with only slight difference. This PR introduces a data type - the **DataTreeDitto** that behaves very similarly to xarray DataTrees with slight differences.

DataTrees are xarrays hierarchical data type:

<img width="960" height="796" alt="image" src="https://github.com/user-attachments/assets/6ccbf2a4-2719-4d80-a0fa-3faf42a6667a" />

---

## ✅ Checklist for Reviewer Reference (filled out by PR Author)

- [x] **Tests**: All tests and CI pass (e.g., full, base, extra, etc.)
    - [x] **Full test**: Yes, full test *is passing*.
- [x] **Integration Tests**: Integration tests added/updated for new/modified functionality *OR* no new functionality
- [x] **Other Repos**: I have updated other repositories to handle this change *OR* my package doesn't impact other plugin packages.

For more details, please see our [review template](https://github.com/NRLMMD-GEOIPS/.github/blob/main/.github/review-template.md) 💌

---

## 🧪 Testing Instructions

This PR comes with 199 associated tests. I wrote 40 of them and then the other ~160 or so are from xarray. I grabbed the unit tests that xarray uses to test their own DataTree implementation to ensure this is compatible with native DataTrees. 

<img width="325" height="38" alt="Screenshot 2025-09-29 at 3 08 52 PM" src="https://github.com/user-attachments/assets/64c73b90-adee-4c06-b5df-7ea5ccdb901c" />

Please run the unit tests. 

It also adds new documentation. Please read over that.

Note: some unit tests from xarray's latest are marked as xfail as they test functionality that hasn't been released yet. This currently matches the latest version of xarray so be sure to update.

💖🌟🌎🌟💖

biosafetylvl5 committed:
Create DataTreeDitto.py

biosafetylvl5 committed:
Update Dockerfile - fix system vs site mismatch

biosafetylvl5 committed:
Update Dockerfile to add dev dependencies

biosafetylvl5 committed:
Merge branch 'biosafetylvl5-patch-1' into datatree-ditto

biosafetylvl5 committed:
First go at implementing DataTreeDitto

biosafetylvl5 committed:
Update .gitignore to ignore ruff configs

biosafetylvl5 committed:
Storing changes for now

biosafetylvl5 committed:
Add xarray tests, bug fixes and documentation

biosafetylvl5 committed:
update index files

biosafetylvl5 committed:
Merge branch 'main' into datatree-ditto

biosafetylvl5 committed:
Add blank release note

jsolbrig commented on docs/source/concepts/architecture/custom-types/datatree-ditto.rst:
```diff
@@ -0,0 +1,150 @@
+
+Define conversion functions:
+
+.. code-block:: python
+
+   def list_to_dataset(lst, name="data", **kwargs):
+       return DataTreeDitto._numpy_to_dataset(np.array(lst), name=name)
+
+   def dataset_to_list(ds: xr.Dataset):
+       return ds[list(ds.data_vars)[0]].values.tolist()
+
+Register the converter:
+
+.. code-block:: python
+
+   DataTreeDitto.register_converter(list, list_to_dataset, dataset_to_list)
```
How would this handle something more complex? Two examples come to mind: lists of numpy arrays and dictionaries of xarray datasets.

biosafetylvl5 replied:
You would write a converter that takes the input type and returns something supported natively by xarray datatrees - for xarray datasets those are natively supported so no converter is needed.

jsolbrig commented on docs/source/concepts/architecture/custom-types/datatree-ditto.rst:
```diff
@@ -0,0 +1,150 @@
+
+.. code-block:: python
+
+   DataTreeDitto.register_converter(list, list_to_dataset, dataset_to_list)
```
Is the first argument here the type that is intended to be handled by the converter?

jsolbrig replied:
Yep!

jsolbrig commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+                return DataTreeDitto._convert_datatree_to_ditto(result)
+            return result
+
+        return f_that_returns_datatree_ditto
```
Can this be commented since it gets used all over the place? 

It appears that its job is to consistently convert DataTree instances back to DataTreeDitto after calling DataTree methods. Is that correct?

jsolbrig replied:
Reading this more, it appears that what this does is:
- If the resulting data are not a DataTree => return the data
- If the resulting data ***area*** a DataTree => return DataTreeDitto

Is that correct? Decorators always wrap my brain in knots a bit so I just want to be sure that I'm understanding.

Maybe the decorator could be renamed a bit to be clearer? Something like `_if_output_DataTree_covert_to_DataTreeDitto` but preferably shorter? I'm suggesting this because it's not clear from the decorator name that it will only convert DataTrees and will leave other object types alone and it would be nice to know that in the context where the decorator is used.

biosafetylvl5 replied:
This is a bug; I added a type error on not DT-like types

jsolbrig commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+        new_ditto = DataTreeDitto(dataset=dt.ds, name=dt.name)
+        for child_name, child in dt.children.items():
+            new_ditto[child_name] = DataTreeDitto._convert_datatree_to_ditto(child)
+        return new_ditto
```
How expensive is this? It's only using copies of data I assume?

jsolbrig commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+        Registers converters for numpy arrays automatically when the class
+        is first instantiated.
+        """
+        cls.register_converter(np.ndarray, cls._numpy_to_dataset, cls._dataset_to_numpy)
```
So, the built-in converters will be added directly to this class, correct?

jsolbrig commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+        cls._converters[obj_type] = {
+            "to_dataset": to_dataset_func,
+            "from_dataset": from_dataset_func,
+        }
```
Given that these are registered by their data type, will it correctly handle data types that are subclasses of registered data types?

I think this is actually two questions:
1. If both `numpy.array` and `numpy.masked_array` are registered, would DataTreeDitto always select the correct type?
2. If only `numpy.array` is registered, would DataTreeDitto attempt to handle `numpy.masked_array` since it is a subclass of `numpy.array` or would it error because `numpy.masked_array` isn't registered?

biosafetylvl5 replied:
1. No, this is a current limitation of this approach using duck-typing.
2. Yes

As we talked about, it is likely that we could solve this by: type checking more aggressively OR creating converters that handle types and their children. In this example, it would be a converter that handles numpy arrays and masked arrays.

jsolbrig reviewed (changes requested):
Not done yet but here are some of my comments.

jsolbrig commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+        original_type = dataset.attrs.get("_ditto_original_type")
+
+        if original_type is None:
+            return dataset
+
+        # Find converter by original type string
+        for obj_type, converter_dict in self._converters.items():
+            if f"{obj_type.__module__}.{obj_type.__name__}" == original_type:
+                return converter_dict["from_dataset"](dataset)
+
+        # Fallback: return dataset if converter not found
+        return dataset
+
+    def _intercept_assignment(func):
+        """Decorator to intercept and convert assignments.
+
+        Parameters
+        ----------
+        func : callable
+            Function to wrap (typically __setitem__).
+
+        Returns
+        -------
+        callable
+            Wrapped function that automatically converts assigned values.
+        """
```
I think I could use a bit of an explanation on what this function is actually doing. Or, if you can comment it to the point that I can understand a little better that would be even better.

biosafetylvl5 replied:
I added more comments + updates the docstring

jsolbrig commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+        return new_ditto
+
+    def get_original(self, path: str = ".") -> Any:
+        """Get the original object (before conversion) at the specified path.
```
This would get a new object that looks similar to the original but with potentially modified data, correct?

biosafetylvl5 replied:
This returns a functionally original object HOWEVER that "same" object has all of the same caveats of the original object. Namely if you modify a shallow copy the "original" data will also be modified.

biosafetylvl5 committed:
Merge branch 'main' into datatree-ditto

biosafetylvl5 committed:
Merge branch 'datatree-spec' into datatree-ditto

biosafetylvl5 commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+        name : str, default "data"
+            Name for the data variable in the resulting dataset.
+        dims : list of str, optional
+            Dimension names. If None, generates names like "dim_0", "dim_1", etc.
```
I added more details to the docstring

biosafetylvl5 commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+        if dataset is not None:
+            # if isinstance(dataset, xr.DataArray):
+            #    # Convert DataArray to Dataset
+            #    dataset = dataset.to_dataset()
```
It was implemented later in xarray natively :)

biosafetylvl5 commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+                ),
+            ):
+                # Convert non-xarray object
+                dataset = self._convert_to_dataset(dataset)
```
that's right - I added more documentation to this section

biosafetylvl5 commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+
+    def __init__(self, dataset=None, children=None, name=None):
+        """Initialize a DataTreeDitto instance.
+
+        Parameters
+        ----------
+        data : array-like, xarray.Dataset, xarray.DataArray, or None, optional
+            Data to store in this node. Non-xarray objects are automatically
+            converted to xarray.Dataset using registered converters.
+        parent : DataTreeDitto, optional
+            Parent node in the tree structure.
+        children : dict, optional
+            Dictionary of child nodes.
+        name : str, optional
+            Name of this node.
```
I fixed this

biosafetylvl5 commented on docs/source/concepts/architecture/custom-types/datatree-ditto.rst:
```diff
@@ -0,0 +1,150 @@
+    >>> dt.ds.data.values.tolist()
+    [[1, 2], [3, 4]]
+    >>> dt.get_original().tolist()
+    [1, 2, 3]
```
Yes, that's incorrect - fixed! We should probably be using pydocstring tests...

biosafetylvl5 commented on docs/source/concepts/architecture/custom-types/datatree-ditto.rst:
```diff
@@ -0,0 +1,150 @@
+Usage Example::
+
+    >>> import numpy as np
+    >>> from your_module import DataTreeDitto
```
Probably the right position in GeoIPS.

biosafetylvl5 commented on docs/source/concepts/architecture/custom-types/datatree-ditto.rst:
```diff
@@ -0,0 +1,150 @@
+    [1, 2, 3]
+
+DataTreeDitto internally relies on xarray's DataTrees and metadata used for round-trip conversion of non-xobjects is
+stored in dataset attributes prefixed with ``_ditto_``.
```
Fixed - thanks

biosafetylvl5 committed:
update branch according to PR review from jeremy

biosafetylvl5 committed:
Merge branch 'datatree-spec' into datatree-ditto

mindyls commented on docs/source/concepts/architecture/custom-types/datatree-ditto.rst:
```diff
@@ -0,0 +1,199 @@
+
+When converting an object, ``DataTreeDitto`` first checks for an exact type
+match. If none is found, it falls back to ``isinstance``-based matching,
+selecting the most specific registered type (deepest in the object's MRO).
```
Define mro

jsolbrig replied:
MRO is a python concept. It is the Method Resolution Order for objects. This defines the order in which functions are resolved from parent classes when something like `super()` is called. I'd say this is something that can be looked up.

biosafetylvl5 replied:
It's also not just Python, it's more of a general CS concept

jsolbrig replied:
TIL!

mindyls commented on docs/source/releases/latest/datatree-ditto.yaml:
```diff
@@ -0,0 +1,126 @@
+bug fix:
```
No contents

biosafetylvl5 replied:
oh no

biosafetylvl5 replied:
updated

mindyls commented on tests/unit_tests/utils/types/test_datatree_ditto_FROM_XARRAY.py:
```diff
@@ -0,0 +1,2760 @@
+ON_WINDOWS = sys.platform == "win32"
+
+
+@pytest.fixture(scope="module")
```
Are all the tests in this file literally copy pasted from xarray? Is there some way to just reference the tests in xarray rather than copy pasting thousands of lines of tests that may change?

At least mark where the explicit copy pasting starts / stops vs where geoips specific lines exist?

biosafetylvl5 replied:
:( I could not find a way to just reference them - all of the FROM_XARRAY tests are from XARRAY

evrose54 committed:
Merge branch 'datatree-spec' into datatree-ditto

biosafetylvl5 committed:
Merge branch 'datatree-spec' into datatree-ditto

mindyls commented on geoips/utils/types/datatree_ditto.py:
```diff
@@ -0,0 +1,600 @@
+        converter = self._converters[obj_type]["to_dataset"]
+        return converter(obj, **kwargs)
+
+    def _convert_from_dataset(self, dataset: xr.Dataset) -> Any:
```
I expect a lot of algorithms do not have the same type for input and output ... I was envisioning datatree ditto would handle conversions both on the input and output ?

jsolbrig replied:
We will be moving this direction, yes. We've been talking more about how this works recently and our direction has changed some. I think we will merge this, then @biosafetylvl5 will add a follow-on PR that updates to handle more data types and allow different input and output data types.

evrose54 committed:
Merge branch 'datatree-spec' into datatree-ditto

biosafetylvl5 committed:
Merge branch 'datatree-spec' into datatree-ditto

jsolbrig committed:
Merge branch 'datatree-spec' into datatree-ditto

biosafetylvl5 committed:
Merge branch 'datatree-spec' into datatree-ditto

biosafetylvl5 committed:
Merge branch 'main' into datatree-ditto

biosafetylvl5 committed:
Add DataTreeDitto data type with detailed description

Introduces the DataTreeDitto data type, which mimics xarray DataTrees but allows for non-xobjects, providing similar benefits with some computational overhead. This change includes a detailed description of the DataTreeDitto's functionality and its intended future use.

biosafetylvl5 committed:
linting...

biosafetylvl5 committed:
update release note
